### ISSUE TITLE
How to measure shared memory transport's latency?

### ISSUE THREAD

**Issue by liyue75** [eProsima/Fast-DDS#1590]
I want to measure the latency of SHM transport.
And what I did is to modify the code in examples/C++/HelloWorldExampleSharedMem/HelloWorldPublisher.cpp and HelloWorldSubscriber.cpp:
In publisher.cpp : I use gettimeofday() to get the system time in publish function, then write the message, and then printf the time.
In subscriber.cpp: I also get the system time by gettimeofday in onNewDataMessage function, then printf the time.
then I built the example, and ran pub and sub in two terminals. The publisher would pub ten messages, and the sub recv them.
And I calculate the latency by the time they print.

## Expected Behavior
It is expected that Shared Memory latency should be subtle level.

## Current Behavior
I found the average latency was about 2 milliseconds

## System information
ubuntu20.04

It doesn't make sense,  did I do something wrong or did I have any options that I didn't set?

**Comment by EduPonz** [comment]
@liyue75 , thanks for your report!

We have some performance tests tools already available that will let you toy with measuring latency performance for SHM. Given that you are building the library with `colcon`, you can build the performance tests by:
```
colcon build --packages-select fastrtps --cmake-args -DGTEST_INDIVIDUAL=ON -DPERFORMANCE_TESTS=ON
```

Then, you can run the SHM latency tests with:
```
colcon test --packages-select fastrtps --event-handlers=console_direct+ --ctest-args -R "latency.*shm.*"
```

This will run SHM latency tests using best-effort and reliable reliability. In addition, if you compiled with `-DSECURITY=ON`, you'll have tests for best-effort + security, and for reliable + security.

By default, these tests output latency performance for 16, 1024, and 16384 bytes of message size. However, you can measure for as many sizes as you want by modifying the [payload_demands.csv](https://github.com/eProsima/Fast-DDS/blob/master/test/performance/latency/payloads_demands.csv) file and re-running the tests. These tests produce CSV output files that you can find in `build/fastrtps/test/performance/latency/measurements_*.csv`

I just run some of them locally, and the performance is in the range of tens of micro-seconds (even though my library is compiled with symbols and other inspection tools).

I hope that helps!

**Comment by liyue75** [comment]
@EduPonz Thanks a lot.
Now I use the test tool and get the latency again. 
By change message size, I find the latency is affected by the data size.
The average latency of different packet sizes in reliable shm mode is:
16B  71us
1K    71us
16K  73us
1M  1012us
Now I know why the latency I tested is 2ms, because in example HelloWorldExampleSharedMem the default message size is 2M.
It seems that in the case of large data, the latency of Shared Mem is a high.

**Comment by EduPonz** [comment]
Hi @liyue75 

SHM transport is, as of now, still a transport in the sense that, the sending and receiving buffers, even though they are located in a shared piece of memory, are separated.
This means that when using SHM between two different processes, each `write` operation entails 4 copies of the data, them being:

1. From the writer (user) to the writer's history
1. From the writer's history to the shared memory space
1. From the shared memory space to the reader's history
1. From the reader's history to the reader (user)

Knowing this, it makes sense that, as the data sizes get larger, the latency increases, since the copy operations take more time. However, we are in the progress of enabling a true zero-copy SHM transport, meaning that:

1. The reader and writer will share the data buffer
1. The writer will be able to directly write in the data buffer
1. The reader will read directly from the data buffer

With this mechanism in place, all the overhead caused by the data copying disappears, thus drastically reducing latency. This zero-copy SHM transport will be shipped in the upcoming v2.2.0, which is schedule for the end of the year, stay tuned!

In any case, you can still improve the latency performance by tuning the transport to your specific use case. For that, you'd need use your custom SHM transport.
More specifically, you'd be interested in:
1. Setting the `maxMessageSize` to the size of your data to avoid fragmentation.
1. Adjusting the `segment_size` to be able to accommodate some full pieces data.

You can do both of these things using [XML configuration files](https://fast-dds.docs.eprosima.com/en/latest/fastdds/xml_configuration/transports.html#transport-descriptors)

**Comment by liyue75** [comment]
@EduPonz Thanks for your reply.
